ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/Unit.hs

```diff
@@ -82,13 +100,84 @@ ouroboros_network_4183 =
     f <- newFollower
     void $ followerForward f [blockPoint b1]
     void $ addBlock $ mkNextBlock b4 4 $ fork 0
-    persistBlks
+    persistBlks DoNotGarbageCollect
     void $ addBlock $ mkNextBlock b3 3 $ fork 1
     followerInstruction f >>= \case
-      Right (Just (RollBack _actual))
-        -> assertEqual (blockPoint b1) _actual "Rollback to wrong point"
+      Right (Just (RollBack actual))
+        -> assertEqual (blockPoint b1) actual "Rollback to wrong point"
       _ -> failWith "Expecting a rollback"
 
+
+ouroboros_network_3999 :: (
+    Mock.HasHeader (Block m)
+  , Block m ~ TestBlock
+  , SupportsUnitTest m
+  , MonadError TestFailure m
+  ) => m ()
+ouroboros_network_3999 = do
+    b1 <- addBlock $ firstBlock 0 $ fork 1
+    b2 <- addBlock $ mkNextBlock b1 1 $ fork 1
+    b3 <- addBlock $ mkNextBlock b2 2 $ fork 1
+    i <- streamAssertSuccess (inclusiveFrom b1) (inclusiveTo b3)
+    b4 <- addBlock $ mkNextBlock b1 3 $ fork 2
+    b5 <- addBlock $ mkNextBlock b4 4 $ fork 2
+    b6 <- addBlock $ mkNextBlock b5 5 $ fork 2
+    void $ addBlock $ mkNextBlock b6 6 $ fork 2
+    persistBlks GarbageCollect
+
+    -- b1 is part of the current chain, so should always be streamed
+    iteratorNext i >>= hasResult b1
+
+    -- the remainder of the range of the iterator consists of blocks that may
+    -- have been garbage collected.
+    let expected = [blockRealPoint b2, blockRealPoint b3]
+    remainder <- allResults (length expected + 1) i
+    remainder `isPrefixOf` expected
+      `orFailWith` "Iterator should return part of the prespecified interval"
+
+  where
+    fork i = TestBody i True
+
+    extractBlock (blk, _, _, _, _, _, _, _, _, _, _) = blk
+
+    hasResult expected (API.IteratorResult actual) =
+      assertEqual expected (extractBlock actual) "Wrong block"
+    hasResult _ _ = failWith "Expected IteratorResult"
+
+    -- Bound the number of iterations by `maxIter` to avoid test stalling
+    allResults maxIter it = go maxIter
```

Comment:
It seems to me we know exactly what each iterator should return, it would make the test clearer to explicitly list the result. Or perhaps there's seom form of non-determinism here?

Reply:
There is non-determinism here, since blocks may or may not have been garbage-collected.